File: frontend/Frontend/lib/models/session_models.dart
class Mistake {
  final String word;
  final int ayah;
  final int surah;
  final List<String> tajweedRules;
  final String errorType;
  final double similarity;
  final String occurredAt;
  final String sessionId;

  Mistake({
    required this.word,
    required this.ayah,
    required this.surah,
    this.tajweedRules = const [],
    this.errorType = 'mispronunciation',
    this.similarity = 0.0,
    this.occurredAt = '',
    this.sessionId = '',
  });

  factory Mistake.fromJson(Map<String, dynamic> json) {
    return Mistake(
      word: (json['word'] ?? '').toString(),
      ayah: (json['ayah'] as num?)?.toInt() ?? 0,
      surah: (json['surah'] as num?)?.toInt() ?? 0,
      tajweedRules: (json['tajweedRules'] as List?)?.map((e) => e.toString()).toList() ??
          (json['tajweed_rules'] as List?)?.map((e) => e.toString()).toList() ??
          const [],
      errorType: (json['errorType'] ?? json['error_type'] ?? 'mispronunciation').toString(),
      similarity: (json['similarity'] as num?)?.toDouble() ?? 0.0,
      occurredAt: (json['occurredAt'] ?? json['occurred_at'] ?? '').toString(),
      sessionId: (json['sessionId'] ?? json['session_id'] ?? '').toString(),
    );
  }

  Map<String, dynamic> toJson() => {
        'word': word,
        'ayah': ayah,
        'surah': surah,
        'tajweedRules': tajweedRules,
        'errorType': errorType,
        'similarity': similarity,
        'occurredAt': occurredAt,
        'sessionId': sessionId,
      };
}

class RecitationSession {
  final String id;
  final String userId;
  final int surah;
  final int ayah;
  final String mode;
  final double accuracyScore;
  final double whisperScore;
  final double mfccScore;
  final String dateTime;
  final int durationSeconds;
  final List<Mistake> mistakes;
  final int totalWords;
  final int correctWords;
  final int closeWords;
  final int missingWords;
  final int extraWords;
  final String? referenceAudioUrl;
  final String transcribedText;
  final String correctText;

  RecitationSession({
    required this.id,
    required this.userId,
    required this.surah,
    required this.ayah,
    required this.dateTime,
    this.mode = 'recitation',
    this.accuracyScore = 0.0,
    this.whisperScore = 0.0,
    this.mfccScore = 0.0,
    this.durationSeconds = 0,
    this.mistakes = const [],
    this.totalWords = 0,
    this.correctWords = 0,
    this.closeWords = 0,
    this.missingWords = 0,
    this.extraWords = 0,
    this.referenceAudioUrl,
    this.transcribedText = '',
    this.correctText = '',
  });

  factory RecitationSession.fromJson(Map<String, dynamic> json) {
    return RecitationSession(
      id: (json['id'] ?? '').toString(),
      userId: (json['userId'] ?? json['user_id'] ?? '').toString(),
      surah: (json['surah'] as num?)?.toInt() ?? 0,
      ayah: (json['ayah'] as num?)?.toInt() ?? 0,
      mode: (json['mode'] ?? 'recitation').toString(),
      accuracyScore: (json['accuracyScore'] as num?)?.toDouble() ??
          (json['accuracy_score'] as num?)?.toDouble() ??
          0.0,
      whisperScore: (json['whisperScore'] as num?)?.toDouble() ??
          (json['whisper_score'] as num?)?.toDouble() ??
          0.0,
      mfccScore: (json['mfccScore'] as num?)?.toDouble() ??
          (json['mfcc_score'] as num?)?.toDouble() ??
          0.0,
      dateTime: (json['dateTime'] ?? json['date_time'] ?? '').toString(),
      durationSeconds: (json['durationSeconds'] as num?)?.toInt() ??
          (json['duration_seconds'] as num?)?.toInt() ??
          0,
      mistakes: (json['mistakes'] as List?)
              ?.map((e) => Mistake.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
      totalWords: (json['totalWords'] as num?)?.toInt() ??
          (json['total_words'] as num?)?.toInt() ??
          0,
      correctWords: (json['correctWords'] as num?)?.toInt() ??
          (json['correct_words'] as num?)?.toInt() ??
          0,
      closeWords: (json['closeWords'] as num?)?.toInt() ??
          (json['close_words'] as num?)?.toInt() ??
          0,
      missingWords: (json['missingWords'] as num?)?.toInt() ??
          (json['missing_words'] as num?)?.toInt() ??
          0,
      extraWords: (json['extraWords'] as num?)?.toInt() ??
          (json['extra_words'] as num?)?.toInt() ??
          0,
      referenceAudioUrl:
          (json['referenceAudioUrl'] ?? json['reference_audio_url'])?.toString(),
      transcribedText: (json['transcribedText'] ?? json['transcribed_text'] ?? '').toString(),
      correctText: (json['correctText'] ?? json['correct_text'] ?? '').toString(),
    );
  }

  int get daysAgo {
    try {
      return DateTime.now().difference(DateTime.parse(dateTime).toLocal()).inDays;
    } catch (_) {
      return 0;
    }
  }
}

class DailyActivity {
  final String day;
  final String date;
  final bool hasSession;
  final int sessionCount;
  final int totalMinutes;
  final double avgAccuracy;

  DailyActivity({
    required this.day,
    required this.date,
    required this.hasSession,
    this.sessionCount = 0,
    this.totalMinutes = 0,
    this.avgAccuracy = 0.0,
  });

  factory DailyActivity.fromJson(Map<String, dynamic> json) => DailyActivity(
        day: (json['day'] ?? '').toString(),
        date: (json['date'] ?? '').toString(),
        hasSession: json['hasSession'] == true || json['has_session'] == true,
        sessionCount: (json['sessionCount'] as num?)?.toInt() ??
            (json['session_count'] as num?)?.toInt() ??
            0,
        totalMinutes: (json['totalMinutes'] as num?)?.toInt() ??
            (json['total_minutes'] as num?)?.toInt() ??
            0,
        avgAccuracy: (json['avgAccuracy'] as num?)?.toDouble() ??
            (json['avg_accuracy'] as num?)?.toDouble() ??
            0.0,
      );
}

class RecentRecitation {
  final String title;
  final String mode;
  final double accuracy;
  final int daysAgo;
  final String sessionId;
  final String dateTime;

  RecentRecitation({
    required this.title,
    required this.mode,
    required this.accuracy,
    required this.daysAgo,
    required this.sessionId,
    required this.dateTime,
  });

  factory RecentRecitation.fromJson(Map<String, dynamic> json) => RecentRecitation(
        title: (json['title'] ?? '').toString(),
        mode: (json['mode'] ?? 'recitation').toString(),
        accuracy: (json['accuracy'] as num?)?.toDouble() ?? 0.0,
        daysAgo: (json['daysAgo'] as num?)?.toInt() ??
            (json['days_ago'] as num?)?.toInt() ??
            0,
        sessionId: (json['sessionId'] ?? json['session_id'] ?? '').toString(),
        dateTime: (json['dateTime'] ?? json['date_time'] ?? '').toString(),
      );

  String get timeAgoDisplay {
    if (daysAgo <= 0) return 'Today';
    if (daysAgo == 1) return '1 day ago';
    return '$daysAgo days ago';
  }
}

class WeeklyProgressSummary {
  final int thisWeekCount;
  final double avgAccuracy;
  final int perfectCount;
  final List<DailyActivity> days;
  final List<RecentRecitation> recentRecitations;

  WeeklyProgressSummary({
    required this.thisWeekCount,
    required this.avgAccuracy,
    required this.perfectCount,
    this.days = const [],
    this.recentRecitations = const [],
  });

  factory WeeklyProgressSummary.fromJson(Map<String, dynamic> json) => WeeklyProgressSummary(
        thisWeekCount: (json['thisWeekCount'] as num?)?.toInt() ?? 0,
        avgAccuracy: (json['avgAccuracy'] as num?)?.toDouble() ?? 0.0,
        perfectCount: (json['perfectCount'] as num?)?.toInt() ?? 0,
        days: (json['days'] as List?)
                ?.map((e) => DailyActivity.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
        recentRecitations: (json['recentRecitations'] as List?)
                ?.map((e) => RecentRecitation.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
      );
}

class MistakeWord {
  final String word;
  final int times;
  final double lastSimilarity;
  final String lastOccurredAt;

  MistakeWord({
    required this.word,
    required this.times,
    required this.lastSimilarity,
    required this.lastOccurredAt,
  });

  factory MistakeWord.fromJson(Map<String, dynamic> json) => MistakeWord(
        word: (json['word'] ?? '').toString(),
        times: (json['times'] as num?)?.toInt() ?? 0,
        lastSimilarity: (json['lastSimilarity'] as num?)?.toDouble() ??
            (json['last_similarity'] as num?)?.toDouble() ??
            0.0,
        lastOccurredAt: (json['lastOccurredAt'] ?? json['last_occurred_at'] ?? '').toString(),
      );
}

class AyahMistakeSummary {
  final int surahNumber;
  final String surahName;
  final int ayah;
  final int mistakeCount;
  final String lastPracticedAt;
  final List<MistakeWord> words;

  AyahMistakeSummary({
    required this.surahNumber,
    required this.surahName,
    required this.ayah,
    required this.mistakeCount,
    required this.lastPracticedAt,
    this.words = const [],
  });

  factory AyahMistakeSummary.fromJson(Map<String, dynamic> json) => AyahMistakeSummary(
        surahNumber: (json['surahNumber'] as num?)?.toInt() ??
            (json['surah_number'] as num?)?.toInt() ??
            0,
        surahName: (json['surahName'] ?? json['surah_name'] ?? '').toString(),
        ayah: (json['ayah'] as num?)?.toInt() ?? 0,
        mistakeCount: (json['mistakeCount'] as num?)?.toInt() ??
            (json['mistake_count'] as num?)?.toInt() ??
            0,
        lastPracticedAt: (json['lastPracticedAt'] ?? json['last_practiced_at'] ?? '').toString(),
        words: (json['words'] as List?)
                ?.map((e) => MistakeWord.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
      );
}

class MistakesSummary {
  final List<AyahMistakeSummary> byAyah;
  final int totalMistakes;
  final String? mostRecentMistakeAt;

  MistakesSummary({
    this.byAyah = const [],
    this.totalMistakes = 0,
    this.mostRecentMistakeAt,
  });

  factory MistakesSummary.fromJson(Map<String, dynamic> json) => MistakesSummary(
        byAyah: (json['byAyah'] as List?)
                ?.map((e) => AyahMistakeSummary.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
        totalMistakes: (json['totalMistakes'] as num?)?.toInt() ?? 0,
        mostRecentMistakeAt: (json['mostRecentMistakeAt'])?.toString(),
      );
}

class HomeMetricsWithStreak {
  final int currentStreak;
  final int longestStreak;
  final String? lastSessionDate;
  final int todayMinutes;
  final int thisWeekMinutes;

  HomeMetricsWithStreak({
    required this.currentStreak,
    required this.longestStreak,
    this.lastSessionDate,
    this.todayMinutes = 0,
    this.thisWeekMinutes = 0,
  });

  factory HomeMetricsWithStreak.fromJson(Map<String, dynamic> json) =>
      HomeMetricsWithStreak(
        currentStreak: (json['currentStreak'] as num?)?.toInt() ?? 0,
        longestStreak: (json['longestStreak'] as num?)?.toInt() ?? 0,
        lastSessionDate: json['lastSessionDate']?.toString(),
        todayMinutes: (json['todayMinutes'] as num?)?.toInt() ?? 0,
        thisWeekMinutes: (json['thisWeekMinutes'] as num?)?.toInt() ?? 0,
      );
}

File: backend/session_models.py
# -*- coding: utf-8 -*-
"""
Session and Mistake Data Models for ReciteRight
Tracks recitation sessions and word-level mistakes
"""
from dataclasses import dataclass, field, asdict
from typing import List, Optional, Dict, Any
from datetime import datetime
import json


@dataclass
class Mistake:
    """Word-level mistake from a recitation session"""
    word: str  # Arabic word
    ayah: int
    surah: int
    tajweed_rules: List[str] = field(default_factory=list)  # e.g., ["Madd", "Ghunnah"]
    error_type: str = "mispronunciation"  # "mispronunciation", "skipped", "extra", "timing"
    similarity: float = 0.0  # 0-1, how close was the user to correct
    occurred_at: str = ""  # ISO timestamp
    session_id: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Mistake':
        return cls(**data)


@dataclass
class RecitationSession:
    """Complete recitation session with all metrics and mistakes"""
    id: str
    user_id: str
    surah: int
    ayah: int
    mode: str  # "recitation", "lesson", "review"
    accuracy_score: float  # 0-100 (final_score from /api/compare)
    whisper_score: float  # 0-100 (transcription accuracy)
    mfcc_score: float  # 0-100 (audio quality)
    date_time: str  # ISO timestamp
    duration_seconds: int
    mistakes: List[Mistake] = field(default_factory=list)
    total_words: int = 0
    correct_words: int = 0
    close_words: int = 0
    missing_words: int = 0
    extra_words: int = 0
    reference_audio_url: Optional[str] = None
    transcribed_text: str = ""
    correct_text: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "user_id": self.user_id,
            "surah": self.surah,
            "ayah": self.ayah,
            "mode": self.mode,
            "accuracy_score": self.accuracy_score,
            "whisper_score": self.whisper_score,
            "mfcc_score": self.mfcc_score,
            "date_time": self.date_time,
            "duration_seconds": self.duration_seconds,
            "mistakes": [m.to_dict() for m in self.mistakes],
            "total_words": self.total_words,
            "correct_words": self.correct_words,
            "close_words": self.close_words,
            "missing_words": self.missing_words,
            "extra_words": self.extra_words,
            "reference_audio_url": self.reference_audio_url,
            "transcribed_text": self.transcribed_text,
            "correct_text": self.correct_text,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'RecitationSession':
        mistakes_data = data.get("mistakes", [])
        mistakes = [Mistake.from_dict(m) for m in mistakes_data]
        data_copy = dict(data)
        data_copy["mistakes"] = mistakes
        return cls(**data_copy)


@dataclass
class DailyActivity:
    """Activity for a single day"""
    day: str  # "Mon", "Tue", etc.
    date: str  # YYYY-MM-DD
    has_session: bool
    session_count: int = 0
    total_minutes: int = 0
    avg_accuracy: float = 0.0

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


@dataclass
class RecentRecitation:
    """Summary of a recent recitation for display"""
    title: str  # e.g., "Surah Al-Baqarah Ayah 5"
    mode: str
    accuracy: float
    days_ago: int
    session_id: str
    date_time: str

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


@dataclass
class WeeklyProgressSummary:
    """Weekly progress overview"""
    this_week_count: int
    avg_accuracy: float
    perfect_count: int  # accuracy >= 95
    days: List[DailyActivity] = field(default_factory=list)
    recent_recitations: List[RecentRecitation] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "thisWeekCount": self.this_week_count,
            "avgAccuracy": self.avg_accuracy,
            "perfectCount": self.perfect_count,
            "days": [d.to_dict() for d in self.days],
            "recentRecitations": [r.to_dict() for r in self.recent_recitations],
        }


@dataclass
class MistakeWord:
    """Single word with mistakes"""
    word: str
    times: int  # How many times this word was mistaken
    last_similarity: float
    last_occurred_at: str

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


@dataclass
class AyahMistakeSummary:
    """Summary of mistakes for an ayah"""
    surah_number: int
    surah_name: str
    ayah: int
    mistake_count: int
    last_practiced_at: str
    words: List[MistakeWord] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "surahNumber": self.surah_number,
            "surahName": self.surah_name,
            "ayah": self.ayah,
            "mistakeCount": self.mistake_count,
            "lastPracticedAt": self.last_practiced_at,
            "words": [w.to_dict() for w in self.words],
        }


@dataclass
class MistakesSummary:
    """Aggregated mistakes summary"""
    by_ayah: List[AyahMistakeSummary] = field(default_factory=list)
    total_mistakes: int = 0
    most_recent_mistake_at: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "byAyah": [a.to_dict() for a in self.by_ayah],
            "totalMistakes": self.total_mistakes,
            "mostRecentMistakeAt": self.most_recent_mistake_at,
        }


@dataclass
class HomeMetricsWithStreak:
    """Home screen metrics including streak"""
    current_streak: int
    longest_streak: int
    last_session_date: Optional[str] = None
    today_minutes: int = 0
    this_week_minutes: int = 0

    def to_dict(self) -> Dict[str, Any]:
        return {
            "currentStreak": self.current_streak,
            "longestStreak": self.longest_streak,
            "lastSessionDate": self.last_session_date,
            "todayMinutes": self.today_minutes,
            "thisWeekMinutes": self.this_week_minutes,
        }

